### ISSUE TITLE
Unet decoder upsampling

### ISSUE THREAD

**Issue by DamienLopez1** [qubvel/segmentation_models.pytorch#187]
Hi

I am using a Unet model with the encoder set to 'resnet34' and the pretrained weights are imagenet.

When I look at the model I do not see where the upsampling is occuring. The convolutions in the encoder side are occuring (although the downsampling is seemingly  occuring after the intended layer e.g downsampling from layer 1 to layer 2 only occurs after layer 2), however I do not see where the upsampling takes place in the decoder side. 

There is also the case where I do not see the centre block convolutions occuring. 

Can I please be explained where the upsampling occurs?

My model for reference:

[resnet34 Unet model.txt](https://github.com/qubvel/segmentation_models.pytorch/files/4517379/resnet34.Unet.model.txt)

**Comment by howard-mahe** [comment]
Hello, in UNet implementation, the upsampling operator is a nearest neighbor interpolation which uses the PyTorch functional api, see [unet/decoder.py#L36](https://github.com/qubvel/segmentation_models.pytorch/blob/a3cc9ac7b3430967414506f37c5f043e1ba3397f/segmentation_models_pytorch/unet/decoder.py#L36)

Since the functional operators are parameter-free, the upsampling operator in UNet does not appear in the `model.modules()`.

**Comment by DamienLopez1** [comment]
Hi Howard 

For semantic segmentation should the centre block be activated? I have not activated it and my results seem to be very good(dice loss of 0.20 and IOU score of 0.767).

With regards to: 
> the outputs spatial resolution of [Conv, Conv, Pool s=2] are e.g. 1/4, 1/4, 1/8
the outputs spatial resolution of [Pool s=2, Conv, Conv] are e.g. 1/4, 1/4, 1/4 => same spatial resolution over the whole layer
 
How do we modify this spatial resolution?

And finally in using the segmentation head, my segmentation head is using an identity upsample rather than nn.UpsamplingBilinear2d. How do I set upsampling > 1? Is it even necessary?

**Comment by JulienMaille** [comment]
@howard-mahe Sorry to hijack the thread, I had a question regarding the downsampling part of the encoder. I'm trying to teach my network to recognize small features that quickly disappear after 2 or 3 downsamplings. In order to improve my IoU I was wondering if it would make sens to increase the feature channel of the first layers of the encoder. Do you have any literature to recommend on this topic?

**Comment by howard-mahe** [comment]
@DamienLopez1
The center block that is used in the original paper can be seen as part of the encoder. The center block is useless when you use deep backbones as encoder (it is always the case in this library) because the deep backbones have enough convolutions at their last `layer` (e.g. 3 `BasicBlock` in the last `layer` of `resnet34`, see [encoders/resnet.py#L87-88](https://github.com/qubvel/segmentation_models.pytorch/blob/a3cc9ac7b3430967414506f37c5f043e1ba3397f/segmentation_models_pytorch/encoders/resnet.py#L87-L88)). So yes, *keep disabled the center block*.

I am not sure I got your question regarding the modification of the spatial resolution. Actually, all the encoders are already defined so that each `stage` decreases by 2 the spatial resolution. The default number of stages is 6 (see [encoders/resnet.py#L46-L54](https://github.com/qubvel/segmentation_models.pytorch/blob/a3cc9ac7b3430967414506f37c5f043e1ba3397f/segmentation_models_pytorch/encoders/resnet.py#L46-L54) for resnets) where the output spatial resolutions are `[1, 1/2, 1/4, 1/8, 1/16, 1/32]`. If you want your encoder to be "less deep", you can control the number of stages using the `encoder_depth` argument. Otherwise, if you want to maintain the spatial resolution while not modifying the depth of the encoder, you can apply a dilation rate (see [`make_dilated`](https://github.com/qubvel/segmentation_models.pytorch/blob/b77cf5aec7a67d44e839b26dbd46572f49f8612e/segmentation_models_pytorch/encoders/_base.py#L35-L41)) on the convolutions of the last stages, only if if the `encoder` supports this feature.

---

Hi @JulienMaille. Because of the pooling operators (or convolution stride=2), indeed some content might disappear from the feature maps after few stages. Deep backbones are designed to capture both local and global information, hence the theoretical receptive field of a resnet is generally above 200x200. If you are interested in small features, I imagine you do not want to increase the receptive field of your network. So first, do not use pooling operators and second do not make the network too deep. In the literature, Wu et al. have applied such principles to VGG to detect manipulations in images:

_Wu, Y., AbdAlmageed, W., & Natarajan, P. (2019). Mantra-net: Manipulation tracing network for detection and localization of image forgeries with anomalous features. In Proceedings of the IEEE Conference on Computer Vision and Pattern Recognition (pp. 9543-9552)._ ([link](http://openaccess.thecvf.com/content_CVPR_2019/papers/Wu_ManTra-Net_Manipulation_Tracing_Network_for_Detection_and_Localization_of_Image_CVPR_2019_paper.pdf)).

Hence the authors maintained the receptive of field of their "fully dilated VGG" at 23x23.

**Comment by DamienLopez1** [comment]
Thanks Howard that helps alot.

Regarding the identity upsampling, this is my models segmentation head:

> (segmentation_head): SegmentationHead(
    (0): Conv2d(16, 1, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (1): Identity()
    (2): Activation(
      (activation): Sigmoid()
    )
  )

The upsampling is set to identity. Why would this be?

**Comment by howard-mahe** [comment]
In `UNet` architecture, the `UnetDecoder` already upsamples the features up to the input image size, but this might not be the case for other decoders. In that particular case, you do not need to upsample the features further. That's why the `upsampling` argument of the `SegmentationHead` (see [base/heads.py](https://github.com/qubvel/segmentation_models.pytorch/blob/a3cc9ac7b3430967414506f37c5f043e1ba3397f/segmentation_models_pytorch/base/heads.py#L5-L11)) is not set in the definition of `UNet` ([unet/model.py](https://github.com/qubvel/segmentation_models.pytorch/blob/master/segmentation_models_pytorch/unet/model.py)). If you look at the segmentation head implementation, you will find this:
```python
upsampling = nn.UpsamplingBilinear2d(scale_factor=upsampling) if upsampling > 1 else nn.Identity()
```
The segmentation head of `UNet` matches the case where `upsampling == 1` , hence `upsampling == nn.Identity()` because we don't need to upsample the features further.

**Comment by DamienLopez1** [comment]
Thanks alot Howard. I get it now!

**Comment by JulienMaille** [comment]
@howard-mahe Right now I'm working with a Unet/Resnet18 reduced to 4 layers (I tried keeping only 3 but it's slightly worse). By pooling operators did you mean maxPool? Is it worth messing with 'replace_stride_with_dilation'?

I've tried FPN, PAN and DeepLab from this repo but they were all outperformed by Unet. Bigger resnet encoders did not help. Right now I'm experimenting with timm-efficiennet-b3/b4 and got ~2% improvement on IoU

**Comment by howard-mahe** [comment]
@JulienMaille By pooling operators I mean any operator with a stride not equal to 1. There are five of them in any resnet in order to reduce the spatial resolution by a factor 1/32 at the end of the encoder. For your very specific usecase, I imagine you probably don't need to reduce the spatial resolution at all (just like in ManTra-Net) in order to maintain a small receptive field. `replace_stride_with_dilation` won't help you because its purpose is to preserve the receptive field. Please note that if you don't reduce the spatial resolution in the encoder, you don't need a decoder at all. In addition, I don't think this is a good idea to reduce drastically the number of convolutions of your network because you need it to be deep enough to build a representation of the data.

Here follows for instance the implementation of the VGG in ManTra-Net:
```python
from torchvision.models.vgg import cfgs, make_layers

# Remove pooling operators from cfgs
cfgs = {name: list(filter(lambda a: a != 'M', cfg)) for name, cfg in cfgs.items()}

vgg11 = make_layers(cfgs['A'])
vgg13 = make_layers(cfgs['B'])
vgg16 = make_layers(cfgs['D'])
vgg19 = make_layers(cfgs['E'])
```

If you have a lot of training data (1M+), you will be able to train such network from scratch. If you don't, well... you will have a hard time because pre-trained deep neural nets are designed with pooling operators which aim at aggregating information in a large receptive field. You cannot use those pre-trained models. Well, I agree that you can at least use the first stage of pre-trained models when their work at plain resolution.

**Comment by DamienLopez1** [comment]
Maybe this should be another issue, but if I can maybe ge tan explanation as to why DICE loss works as a loss function for Unet rather than the Softmax with BCE loss as said in the Unet paper? I am doing a background vs single class segmentation and am using a sigmoid activation. 

I am thinking that because its single class and sigmoid, this is why dice loss is permissable? Am I on the right track?

**Comment by howard-mahe** [comment]
You are right, this should be another issue.

Well, I am not an expert of binary losses. But the point is that `UNet` was designed in 2014. Since then, practitioners have found that DICE loss can be a good substitute or even complementary to BCE Loss for binary segmentation tasks.

Regarding softmax vs. sigmoid: for the binary segmentation tasks, you better go with the sigmoid and predicts one `heat map` while the softmax must be used for multi-class segmentation tasks. Even if the softmax can be used for binary segmentation tasks when the number of classes K is set to 2, but this is a redundant formulation compared to sigmoid.

**Comment by JulienMaille** [comment]
@howard-mahe Do you know a place (forum, reddit, discord, etc.) where good discussion between experts like you and newbies like me can happen? (in french or english, doesn't matter)

**Comment by DamienLopez1** [comment]
Thanks alot Howard, that clears things up for me.
